src/m2n/PointToPointCommunication.cpp

```diff
@@ -245,42 +281,65 @@ void printLocalIndexCountStats(std::map<int, std::vector<int>> const &m)
  *
  * @returns the resulting communication map for rank thisRank
  *
- * The approximate complexity of this function is:
- * \f$ \mathcal{O}(n \log(n) + m \log(n)) \f$
+ * The worst case complexity of the function is:
+ * \f$ \mathcal{O}(p n \log(n) + p 2 (2 n)) \f$
+ *
+ * which is composed of the initial std::sort for each vector and the subsequent
+ * computation of the intersection.
  *
- * * n is the total number of data indices for all ranks in `otherVertexDistribution'
- * * m is the number of local data indices for the current rank in `thisVertexDistribution`
+ * * n is the number of data indices for each vector in `otherVertexDistribution'
+ * * p number of ranks
+ * * Note that n becomes smaller, if we have more ranks.
  *
+ * However, in case our indices are already sorted (which is typically the case),
+ * the complexity boils down to linear complexity
+ * \f$ \mathcal{O}(p n) \f$
  */
 std::map<int, std::vector<int>> buildCommunicationMap(
     // `thisVertexDistribution' is input vertex distribution from this participant.
-    mesh::Mesh::VertexDistribution const &thisVertexDistribution,
+    mesh::Mesh::VertexDistribution &thisVertexDistribution,
     // `otherVertexDistribution' is input vertex distribution from other participant.
-    mesh::Mesh::VertexDistribution const &otherVertexDistribution,
-    int                                   thisRank = utils::IntraComm::getRank())
+    mesh::Mesh::VertexDistribution &otherVertexDistribution,
+    int                             thisRank = utils::IntraComm::getRank())
 {
   auto iterator = thisVertexDistribution.find(thisRank);
   if (iterator == thisVertexDistribution.end())
     return {};
 
-  // Build lookup table from otherIndex -> rank for the otherVertexDistribution
-  const auto lookupIndexRank = [&otherVertexDistribution] {
-    boost::container::flat_multimap<int, int> lookupIndexRank;
-    for (const auto &other : otherVertexDistribution) {
-      for (const auto &otherIndex : other.second) {
-        lookupIndexRank.emplace(otherIndex, other.first);
-      }
-    }
-    return lookupIndexRank;
-  }();
+  std::map<int, std::vector<int>> communicationMap;
+
+  // take advantage that these data structures are in most cases sorted by construction,
+  // i.e., we perform here mostly a safety check and don't perform an actual sorting
+  if (!std::is_sorted(iterator->second.begin(), iterator->second.end())) {
+    std::sort(iterator->second.begin(), iterator->second.end());
+  }
 
-  auto const &indices = iterator->second;
+  // now we iterate over all other vertex distributions to compute the intersection
+  for (auto &other : otherVertexDistribution) {
```

Comment:
We have C++17, so this could be a good opportunity to clean up the below code with structured bindings:

```suggestion
  for (const auto & [rank, vertices] : otherVertexDistribution) {
```